### PR TITLE
feat: [CO-468] Add service backend for GetDomainCert SOAP API

### PR DESCRIPTION
**Pull request by aheeva-yuliya**

### CODE REVIEW

**Review by frisonisland — CHANGES_REQUESTED** [review]
Need tests! :slightly_smiling_face: 
Also when mailbox changes from PR: https://github.com/Zextras/carbonio-mailbox/pull/136 are merged we can set 23.2.0-SNAPSHOT version and the pipeline will build.

File: src/main/java/com/zimbra/cert/GetDomainCert.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+   *     during certificate parsing.
+   */
+  @Override
+  public Element handle(Element request, Map<String, Object> context) throws ServiceException {
+
+    ZimbraSoapContext zsc = getZimbraSoapContext(context);
+
+    Provisioning prov = Provisioning.getInstance();
+    String domainName = request.getAttribute(AdminConstants.A_DOMAIN);
+    Domain domain = prov.get(DomainBy.name, domainName);
```
Comment (keshavbhatt) [inline src/main/java/com/zimbra/cert/GetDomainCert.java]: To keep the APIs(GetCert and this one) consistent, consider adding the ability to accept requests by domain ID as well. We may accomplish this by making the **name** parameter optional and the **id** parameter obligatory :)

**Review by keshavbhatt — CHANGES_REQUESTED** [review]
Please provide unit tests covering corner cases like:
what happens -
- if certificate is null
- if one of the requested elemente from certificate object is null

File: src/main/java/com/zimbra/cert/GetDomainCert.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+import java.text.DateFormat;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Admin Handler class to get information about a domain certificate. It provides a way to access
+ * all needed attributes of X.509 certificate using a standard java.security.cert package.
+ */
```
Comment (Polpetta) [inline src/main/java/com/zimbra/cert/GetDomainCert.java:28]: Please add the `@since` and `@author` so we know when it was introduced!

**Review by Polpetta — COMMENTED** [review]
I agree with @keshavbhatt suggestion and with the necessity to add tests!

**Comment by aheeva-yuliya** [comment]
> 

@frisonisland @keshavbhatt @Polpetta i started to write some tests 
about certificates and null its ok
but about "checkDomainRight(zsc, domain, Admin.R_getDomain);" its very hard to mock bcz of a lot static metods and final config class 
inside maillbox there is MailboxTestUtil for testing but its in tests and not reacheble from certificare-manager-store
so i just gonna skip this 
the tests about permissions for global and deledated admin should already be in place

File: src/main/java/com/zimbra/cert/GetDomainCert.java (hunk 1 above)
Comment (aheeva-yuliya) [reply src/main/java/com/zimbra/cert/GetDomainCert.java]: i would modify by id then only like it is for a server
i've checked admin panel 
before calling this api they would already have the id information 
no need for additional option by name then

**Comment by keshavbhatt** [comment]
> > 
> 
>  i started to write some tests about certificates and null its ok 

This will be enough in this case, no need to execute full SOAP request.

**Comment by frisonisland** [comment]
> > 
> 
> @frisonisland @keshavbhatt @Polpetta i started to write some tests about certificates and null its ok but about "checkDomainRight(zsc, domain, Admin.R_getDomain);" its very hard to mock bcz of a lot static metods and final config class inside maillbox there is MailboxTestUtil for testing but its in tests and not reacheble from certificare-manager-store so i just gonna skip this the tests about permissions for global and deledated admin should already be in place

Maybe you can take inspiration from these tests: https://github.com/Zextras/carbonio-certificate-manager-store/blob/main/src/test/java/com/zimbra/cert/VerifyCertKeyTest.java
They they simulate the SOAP request and mock some internal process if I'm not mistaken, although it's not an integration test with the server spin up.

Edit: the idea on those tests is: you receive the SOAP request, simulate an authenticated user through context, validate the logic. I think we could do better by splitting the logic from the controller and authentication at service level (just my thought)

File: src/test/java/com/zimbra/cert/GetDomainCertTest.java [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+  @Before
+  public void setUp() throws Exception {
+    final ZimbraSoapContext zsc =
+        new ZimbraSoapContext(mock(AuthToken.class), "1", SoapProtocol.Soap12, SoapProtocol.Soap12);
+    context.put(SoapEngine.ZIMBRA_CONTEXT, zsc);
+    Provisioning.setInstance(provisioning);
+  }
+
+  @Test(expected = ServiceException.class)
```
Comment (frisonisland) [inline src/test/java/com/zimbra/cert/GetDomainCertTest.java]: I think checking only ServiceException is thrown is a bit generic, check also the message/code of the exception is what we want.
You can use: https://junit.org/junit4/javadoc/4.13/org/junit/rules/ExpectedException.html or https://www.baeldung.com/junit-assert-exception

**Comment by keshavbhatt** [comment]
@aheeva-yuliya can you bump the mailbox dependecy version here https://github.com/Zextras/carbonio-certificate-manager-store/blob/feat/CO-486_allow_DA_verify_domain_cert/pom.xml#L9 to make the CI happy? It should work since your merged the mailbox PR.

File: src/test/java/com/zimbra/cert/GetDomainCertTest.java [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+    assertEquals(
+        "<GetDomainCertResponse"
+            + " xmlns=\"urn:zimbraAdmin\">"
+            + "<cert domain=\"domainName\">"
+            + "<subject>CN=webmail-acme.demo.zextras.io</subject>"
+            + "<SubjectAltName>webmail-acme.demo.zextras.io</SubjectAltName>"
+            + "<issuer>CN=R3,O=Let's Encrypt,C=US</issuer>"
+            + "<notBefore>Nov 22 2022 17:26:01 CET</notBefore>"
+            + "<notAfter>Feb 20 2023 17:26:00 CET</notAfter>"
+            + "</cert>"
```
Comment (keshavbhatt) [inline src/test/java/com/zimbra/cert/GetDomainCertTest.java]: This test will fail for timezones other than CET.

File: src/main/java/com/zimbra/cert/GetDomainCert.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+      throw ServiceException.INVALID_REQUEST(
+          "Domain with id " + domainId + " could not be found.", null);
+    }
+
+    if (domain.getSSLCertificate() == null) {
+      throw ServiceException.INVALID_REQUEST(
+          "Certificate for Domain with id " + domainId + " could not be found.", null);
+    }
+
+    checkDomainRight(zsc, domain, Admin.R_getDomain);
```
Comment (keshavbhatt) [inline src/main/java/com/zimbra/cert/GetDomainCert.java]: Use AdminRights base class to refer to the static member R_getDomain.

File: src/test/java/com/zimbra/cert/GetDomainCertTest.java [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+                + "BXrSDRXXrPYnF5rMWhEwDQYJKoZIhvcNAQELBQADggEBAJXIkzjX3thAZyBSj9lr\n"
+                + "It8ipidZNghgI27d7cqntkdsIDWQYyRRbS05GObMoAEiQnFti8EGnqQqEFkgOAj1\n"
+                + "9wXjSQdOGJjzm+giMhci2VMBC4N/7fWOSxWQ+jmVsYOLUFFb8VBuhWpwU5VHOAcv\n"
+                + "6cHuQKNOw8cYm9BFtdKqeexcTkflpWVsNK/CxIiK3cf9/p43MLO73AdGPT2g6DEk\n"
+                + "f+6Ay6pRLopibLZgGWqqBxevG9Pag5VN1CXcR5cu431Ic5l6NLJp58ED9qw8rf/7\n"
+                + "2gAhU3H9TC8UfXXnezzd5PSRjFAi05ciYyUQv+jtAN/U/wGDpiZ1QySvq0p6NC0J\n"
+                + "f+g=\n"
+                + "-----END CERTIFICATE-----\n");
+
+    final GetDomainCert getDomainCert = new DumbGetDomainCertHandler();
```
Comment (frisonisland) [inline src/test/java/com/zimbra/cert/GetDomainCertTest.java:135]: I suggest to find another way to mock this behavior, because:
- we are not using the class that we want to test
- it's not clear what the "dumb" class is effectively doing (I know that checkDomainRight output is ignored, but still). What if we override another method of the class in future?

Consider using Mockito spy() to make it more obvious what we are doing in the test (@Polpetta I know you're not a fan of spying, maybe you can suggest an alternative) or check if it is possible to give the user the Right mocking zcs.

**Review by frisonisland — CHANGES_REQUESTED** [review]

**Review by keshavbhatt — CHANGES_REQUESTED** [review]
I commented few suggestions. Everything else looks good to me.

File: src/test/java/com/zimbra/cert/GetDomainCertTest.java (hunk 6 above)
Comment (frisonisland) [reply src/test/java/com/zimbra/cert/GetDomainCertTest.java:135]: I checked and to mock zcs we would have to go deep downmocking the Provisioning Config class (not worth it).
Also the method is protected, it would have been nice if the Authorization class was provided instead of extending everything. I just don't like wrapping in another class. 
Leave it as is or instead define an Anonymous so we don't encourage using DumbGetDomainCertHandler class.

File: src/test/java/com/zimbra/cert/GetDomainCertTest.java (hunk 3 above)
Comment (aheeva-yuliya) [reply src/test/java/com/zimbra/cert/GetDomainCertTest.java]: lol @frisonisland i've tried import asserThrows but there is no this method there
will check that you propose

Comment (frisonisland) [reply src/test/java/com/zimbra/cert/GetDomainCertTest.java]: > lol @frisonisland i've tried import asserThrows but there is no this method there will check that you propose

Check this: https://github.com/Zextras/carbonio-mailbox/blob/main/store/src/test/java/com/zimbra/cs/service/mail/CopyToFilesIT.java

Comment (aheeva-yuliya) [reply src/test/java/com/zimbra/cert/GetDomainCertTest.java]: @frisonisland yeah i've made with 
@Rule public ExpectedException expectedEx = ExpectedException.none();
 expectedEx.expect(ServiceException.class);
 expectedEx.expectMessage("Domain with id domainId could not be found.");

File: src/test/java/com/zimbra/cert/GetDomainCertTest.java (hunk 6 above)
Comment (aheeva-yuliya) [reply src/test/java/com/zimbra/cert/GetDomainCertTest.java:135]: > Leave it as is or instead define an Anonymous so we don't encourage using DumbGetDomainCertHandler class.

i've made it to be an inner class

**Review by frisonisland — APPROVED** [review]

**Review by keshavbhatt — APPROVED** [review]

**Review by Polpetta — APPROVED** [review]
